test/blackbox-tests/test-cases/build-bin.t

```diff
@@ -0,0 +1,44 @@
+Pform expansion should work in `exec` but also in `build`:
+
+  $ cat > dune-project << EOF
+  > (lang dune 3.10)
+  > (package (name public))
+  > EOF
+  $ cat > dune << EOF
+  > (executable (public_name public) (modules Public))
+  > (executable (name private) (modules Private))
+  > (env
+  >   (_ (binaries private.exe as priv)))
+  > EOF
+
+The executables in question just display what they are:
+
+  $ cat > public.ml << EOF
+  > let () = print_endline "Public"
+  > EOF
+  $ cat > private.ml << EOF
+  > let () = print_endline "Private"
+  > EOF
+
+Making Dune build the executables should work without error messages:
+
+  $ dune build %{bin:public}
+  Error: File unavailable:
+  $TESTCASE_ROOT/../install/default/bin/public
+  [1]
+  $ ls _build/default/public.exe
+  _build/default/public.exe
+
+Building the private executable should also work.
+
+  $ dune build %{bin:private}
+  $ ls _build/default/private.exe
+  _build/default/private.exe
+
+Building the private executable via the env alias should work too:
+
+  $ dune build %{bin:priv}
```

Comment:
By the way, this particular construct should be tested more thoroughly. In particular, `dune build %{bin:priv}` should respect `binaries` in `$CWD`. So, consider the following:

```
$ cat >dune<<EOF
(.. (binaries myexe.exe as bar) ..))
> EOF
$ cat >foo/dune <<EOF
(.. (binaries blah.exe as foo) ..))
> EOF
$ dune build %{bin:bar} # should work
$ dune build %{bin:bar} --root foo/ # should work
$ dune build %{bin:foo} # doesn't work because foo is underneath us
$ dune build %{bin:foo} --root foo/ # should work
```

Reply:
Also note that root discovery is disabled while inside tests. So you will need to use `--root` to set the root manually.